File: Count number of nice subarrays.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
         for(int i = 0; i < nums.length; i++)
            nums[i] = (nums[i] % 2 == 0 ? 0 : 1);
        int result = 0, sum = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        for(int i = 0; i < nums.length; i++){
            sum += nums[i];
            if(map.containsKey(sum - k))
                result += map.get(sum - k);//get how many sub arrays we can generate that has a prefix of sum - k
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
        return result;
    }
}

File: Count Number of Nice Subarrays.py
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        for i in range(len(nums)):
            nums[i] = (0 if nums[i] % 2 == 0 else 1)
        count = defaultdict(int)
        count[0] += 1
        running_sum = result = 0

        for i in range(len(nums)):
            running_sum += nums[i]
            result += count[running_sum - k]
            count[running_sum] += 1

        return result
